File: Laravel/app/Biab/Resources/Storefront.php
<?php

namespace App\Biab\Resources;

use App\Biab\Client;

/** Products, categories, and the per-product detail reads behind a shop. */
class Storefront
{
    public function __construct(private readonly Client $client)
    {
    }

    /** @return array<string, mixed> */
    public function listProducts(?int $limit = null, int|string|null $cursor = null, ?string $categoryId = null): array
    {
        return $this->client->get('storefront/products', [
            'limit' => $limit,
            'cursor' => $cursor,
            'categoryId' => $categoryId,
        ]);
    }

    /**
     * The full shop grid: enriched cards plus `categoryCounts` and the
     * catalog-wide `priceRange` for a filter sidebar.
     *
     * `sort` is one of featured | newest | price-asc | price-desc | rating-desc.
     *
     * @return array<string, mixed>
     */
    public function listProductsWithMeta(
        ?string $search = null,
        ?string $categoryId = null,
        ?int $minPriceCents = null,
        ?int $maxPriceCents = null,
        ?int $minRating = null,
        ?string $sort = null,
        ?int $limit = null,
        int|string|null $cursor = null,
    ): array {
        return $this->client->get('storefront/products', [
            'meta' => '1',
            'search' => $search,
            'categoryId' => $categoryId,
            'minPriceCents' => $minPriceCents,
            'maxPriceCents' => $maxPriceCents,
            'minRating' => $minRating,
            'sort' => $sort,
            'limit' => $limit,
            'cursor' => $cursor,
        ]);
    }

    /** @return array<string, mixed> */
    public function listCategories(): array
    {
        return $this->client->get('storefront/categories');
    }

    /** @return array<string, mixed> */
    public function getProduct(string $productId): array
    {
        return $this->client->get('storefront/products/'.rawurlencode($productId));
    }

    /** @return array<string, mixed> */
    public function getProductReviews(string $productId, ?int $limit = null, int|string|null $cursor = null): array
    {
        return $this->client->get('storefront/products/'.rawurlencode($productId).'/reviews', [
            'limit' => $limit,
            'cursor' => $cursor,
        ]);
    }

    /** @return array<string, mixed> */
    public function getRelatedProducts(string $productId, ?int $limit = null): array
    {
        return $this->client->get('storefront/products/'.rawurlencode($productId).'/related', [
            'limit' => $limit,
        ]);
    }

    /** @return array<string, mixed> */
    public function getProductAddons(string $productId): array
    {
        return $this->client->get('storefront/products/'.rawurlencode($productId).'/addons');
    }
}

File: Phoenix/lib/biab_starter/biab/resources/storefront.ex
defmodule BiabStarter.Biab.Resources.Storefront do
  @moduledoc "Products, categories, and the per-product detail reads behind a shop."

  alias BiabStarter.Biab.Client

  def list(client, opts \\ []) do
    Client.get(client, "storefront/products",
      limit: opts[:limit],
      cursor: opts[:cursor],
      categoryId: opts[:category_id]
    )
  end

  @doc """
  The full shop grid: enriched cards plus `categoryCounts` and the catalog-wide
  `priceRange` for a filter sidebar.

  `sort` is one of featured | newest | price-asc | price-desc | rating-desc.
  """
  def grid(client, params) do
    Client.get(client, "storefront/products",
      meta: "1",
      search: params[:search],
      categoryId: params[:category_id],
      minPriceCents: params[:min_price_cents],
      maxPriceCents: params[:max_price_cents],
      minRating: params[:min_rating],
      sort: params[:sort],
      limit: params[:limit] || 24,
      cursor: params[:cursor]
    )
  end

  def categories(client), do: Client.get(client, "storefront/categories")

  def get(client, id), do: Client.get(client, "storefront/products/#{enc(id)}")

  def related(client, id, limit \\ 4),
    do: Client.get(client, "storefront/products/#{enc(id)}/related", limit: limit)

  def reviews(client, id, limit \\ 5),
    do: Client.get(client, "storefront/products/#{enc(id)}/reviews", limit: limit)

  def addons(client, id), do: Client.get(client, "storefront/products/#{enc(id)}/addons")

  defp enc(v), do: URI.encode(to_string(v), &URI.char_unreserved?/1)
end
